### ISSUE TITLE
ArrayIndexOutOfBoundsException in Mapsforge

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#10128]
A user reported non-reproducible crashes in a scenario where he repeatedly changed from map to a cache, added waypoint with his current position and returned to the map.

His logfile only contains this crash, not sure if it is related (as the time stamp provided by the user does not match the log time stamp (wrong day)).
```
03-03 07:44:39.257 10687 10687 E AndroidRuntime: FATAL EXCEPTION: main
03-03 07:44:39.257 10687 10687 E AndroidRuntime: Process: cgeo.geocaching, PID: 10687
03-03 07:44:39.257 10687 10687 E AndroidRuntime: java.lang.ArrayIndexOutOfBoundsException: length=1874; index=1875
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:385)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:398)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at org.mapsforge.map.layer.Layers.get(Layers.java:336)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at org.mapsforge.map.android.input.TouchGestureHandler.onSingleTapConfirmed(TouchGestureHandler.java:263)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at android.view.GestureDetector$GestureHandler.handleMessage(GestureDetector.java:300)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:105)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:164)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6798)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
03-03 07:44:39.257 10687 10687 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```

```
--- System information ---
Device: ONEPLUS A3003 (OnePlus3, OnePlus)
Android version: 8.0.0
Android build: ONEPLUS A3003_28_180704
c:geo version: 2021.02.14
Google Play services: enabled - 21.06.13 (040400-358943053)
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide caches: own/found disabled archived
Hide waypoints: original visited
HW acceleration: enabled (default state)
System language: da_DK / user-defined language:
System date format: dd/MM/y
Debug mode active: no
System internal c:geo dir: /data/user/0/cgeo.geocaching (21,9 GB free) internal
User storage c:geo dir: /storage/emulated/0/cgeo (21,8 GB free) external non-removable
Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (21,8 GB free) external non-removable
Database: /data/user/0/cgeo.geocaching/databases/data (99,2 MB) on system internal storage
Last backup: 16. jan. 07.19
GPX import path: /storage/emulated/0/cgeo/gpx
GPX export path: /storage/emulated/0/cgeo/gpx
Offline maps path: /storage/emulated/0/cgeo/maps
Map render theme path:
Live map mode: true
Global filter: display all caches
Sailfish OS detected: false
Fine location permission: granted
Write external storage permission: granted
Geocaching sites enabled:
   geocaching.com: Logged in (Login OK) / PREMIUM
Geocaching.com date format: dd/MM/yyyy
Installed c:geo plugins: contacts
BRouter connection available: false
--- End of system information ---
```

**Comment by eddiemuc** [comment]
The crash above looks like a problem in mapsforge. Maybe indexes of layers are not tracked in a thread-safe way and thus many additions/deletions of layers might bring it out of sync (thus the indexoutofbounds).

**Comment by Lineflyer** [comment]
This is one of the "evergreens" in our crash reports....and the last "big one" as it seems.
I will file an issue on mapsforge side...maybe they have an idea.

**Comment by moving-bits** [comment]
Does that mean that we need to synchronize calls like the following: (taken from `NewMap.switchTileLayer`)

```
final Layers layers = this.mapView.getLayerManager().getLayers();
int index = 0;
if (oldLayer != null) {
    index = layers.indexOf(oldLayer.getTileLayer()) + 1;
}
layers.add(index, newLayer.getTileLayer());
```

would become the following?
```
final Layers layers = this.mapView.getLayerManager().getLayers();
int index = 0;
synchronized(layers) {
    if (oldLayer != null) {
        index = layers.indexOf(oldLayer.getTileLayer()) + 1;
    }
    layers.add(index, newLayer.getTileLayer());
}
```

???

(we have a couple of calls to add/remove for layers)

**Comment by moving-bits** [comment]
> Does that mean that we need to synchronize calls like the following: (taken from `NewMap.switchTileLayer`)
> ...
> would become the following?

Does not seem to be necessary. Not only the examples from mapsforge don't do that themselves, but also methods `Layers.add()` etc. are marked as `synchronized`, so that as long as we do not create multiple `LayerManager` objects, all accesses to `add`, `remove` etc. are synchronized.

**Comment by eddiemuc** [comment]
At least part of the weekend is done now, time for some statistics.

Here is the statistic for the ArrayIndexOutOfBoundsException in Mapsforge in October 2021 (by app version):
![image](https://user-images.githubusercontent.com/6909759/139575451-e7981e41-81d2-440e-8b5e-5778aaccbd07.png)

The fix candidate is active in our beta as of 24.10. This maps to apk 20211023.apk. The other beta versions (and productive versions) don't contain the fix. Here are all beta version relevant for October 2021:
![image](https://user-images.githubusercontent.com/6909759/139575566-c030188a-c04a-4974-b8dd-f4a1baccca30.png)
![image](https://user-images.githubusercontent.com/6909759/139575582-75034e84-820e-4bb4-ac2c-aafc6d57dd61.png)

The picture shows us:
* The error is relevant with >1000 occurences in October 2021 alone spread over many different app versions 
* Beta with fix is active now for 7 days without a single occurence of the ArrayOutOfBoundsException
* In constrast in previous two betas (running since 11.10. / 21.10.) the bug occured 53 times 

In summary, I thing this shows that the fix is working. I will reach out to mapsforge with this data.

**Comment by eddiemuc** [comment]
@Lineflyer as time permits I will prepare a PR such that c:geo (release) is using the forked+midified version on cgeo (instead of the one in my own repo). We might have to start with this setup at least for next production release, so I guess it makes sense to do a cleanup here. Hope that's fine. Any idea on when we might release next beta (or even next production) yet?

**Comment by Lineflyer** [comment]
@eddiemuc 
That was the reason I revived and updated that repository, so your PR fits perfectly and we can use that reference until we used the official release to (better) prove to the mapsforge team, that your change solves our issue.

Lets leave this issue open as a reminder to refer again to the mapsforge repo once the change is merged and/or released there.

**Comment by eddiemuc** [comment]
The fix was merged today into mapsforge master. This means that we can switch back to official mapsforge version as soon as their next release comes out.

**Comment by Lineflyer** [comment]
Finally fixed. See #12524
